pkg/controller/jobs/pod/pod_controller.go

```diff
@@ -556,6 +556,34 @@ func (p *Pod) groupTotalCount() (int, error) {
 	return gtc, nil
 }
 
+// podGroupIndex returns the value of GroupIndexLabel for the pod being reconciled at the moment.
+func (p *Pod) podGroupIndex() (*int, error) {
+	if podGroupName(p.pod) == "" {
+		return nil, fmt.Errorf("pod doesn't have a '%s' label", GroupNameLabel)
+	}
+	podGroupTotalCount, err := p.groupTotalCount()
+	if err != nil {
+		return nil, fmt.Errorf("pod group total count is invalid: %v", err)
+	}
+	groupIndex, ok := p.Object().GetAnnotations()[GroupIndexLabel]
+	if !ok {
+		return nil, nil
+	}
+	groupIndexValue, err := strconv.Atoi(groupIndex)
+	if err != nil {
+		return nil, err
+	}
+	if groupIndexValue <= podGroupTotalCount {
+		return nil, fmt.Errorf("incorrect annotation value '%s=%s': group index should be less than group total count",
+			GroupIndexLabel, groupIndex)
+	}
+	if groupIndexValue < 0 {
```

Comment:
A similar pattern repeats also in topology_ungater (and I just realize we don't check this is >=0 there) and probably other places. What about creating a set of helper functions for parsing labels and annotations. I think this could be `ReadUIntFromAnnotation()`, or even `ReadUIntFromAnnotationWithMax`.

Reply:
Good point, I've created an issue for that: https://github.com/kubernetes-sigs/kueue/issues/3651

Reply:
Ok, when we have `ReadUIntFromLabelWithMax` I think it could also be used here.